### PR TITLE
[red-knot] Implement more types in binary and unary expressions

### PR DESCRIPTION
**Pull request by Glyphack**
Implemented some points from https://github.com/astral-sh/ruff/issues/12701

- Handle Unknown and Any in Unary operation
- Handle Boolean in binary operations
- Handle instances in unary operation
- Consider division by False to be division by zero

I thinks it's easier to review these multiple small things in one go. Let me know if it's not to split for next ones.

### CODE REVIEW

File: crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+j = a | b
+
+reveal_type(c)  # revealed: int
+reveal_type(d)  # revealed: int
+reveal_type(e)  # revealed: int
+reveal_type(f)  # revealed: int
+reveal_type(g)  # revealed: int
+reveal_type(h)  # revealed: int
+reveal_type(i)  # revealed: float
+reveal_type(j)  # revealed: bool
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md:24]: nit: this pattern of creating many variables is a legacy holdover from our previous testing infrastructure, which was more limited in the assertions it could make. For cases like this with small expressions and small types, it's more readable to just do e.g. `reveal_type(a + b)  # revealed: int` all in one line. (In fact, a PR changing existing tests to use this style would also be welcome)

Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md:24]: I think we can be more precise in all of these cases (except for division, since we don't have float literals). For example, we know that `True + True == 2`, so that can be `Literal[2]` instead of `int`. In fact I think the best implementation strategy for all these cases (except for `|`, where we can return a `BooleanLiteral`) is to implement the operator for `IntLiteral` and then just have binary operations on `BooleanLiteral` call back in to the same operation, but with `BooleanLiteral[True]` converted to `Literal[1]` and `BooleanLiteral[False]` converted to `Literal[0]`.

File: crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+## Binary operations on booleans
+
+## Basic Arithmetic
+
+```py
+a = True
+b = True
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md:7]: Especially if we go for a bit more precision, as I suggest below, it would be useful to also test with `True` and `False`, or `False` and `False`. We don't need the full matrix, but at least some different variants for different operators. And maybe all four possible combinations for `|`, since that one is special for bools.

File: crates/red_knot_python_semantic/resources/mdtest/unary/instance.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# Unary Operations
+
+```py
+class Number:
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/unary/instance.md:4]: This test is great! Can we also add a test for a class that implements none of these dunders, and assert that we get an "unsupported operation" diagnostic from each operator? (Unless you decide not to implement that in this PR.)

File: crates/red_knot_python_semantic/resources/mdtest/binary/integers.md [hunk 4]
```diff
@@ -37,6 +37,7 @@ b = 2 // 0  # error: "Cannot floor divide object of type `Literal[2]` by zero"
 c = 3 % 0  # error: "Cannot reduce object of type `Literal[3]` modulo zero"
 d = int() / 0  # error: "Cannot divide object of type `int` by zero"
 e = 1.0 / 0  # error: "Cannot divide object of type `float` by zero"
+f = 1 / False  # error: "Cannot divide object of type `Literal[1]` by zero"
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/binary/integers.md:40]: This tests dividing by `False`, which is good, but you also added a line of code below which supports issuing division by zero diagnostic when dividing a bool by zero (i.e. the dividend is a `BooleanLiteral`). Can we add a test for that, too?

And actually also a test for where the dividend is just `bool()`, too!

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 5]
```diff
@@ -543,6 +543,7 @@ impl<'db> TypeInferenceBuilder<'db> {
             Type::Instance(cls)
                 if cls.is_known(self.db, KnownClass::Float)
                     || cls.is_known(self.db, KnownClass::Int) => {}
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:545]: We can also add `KnownClass::Bool` here

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 6]
```diff
@@ -2469,7 +2472,13 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
-
+            (_, Type::Any) => Type::Any,
+            (_, Type::Unknown) => Type::Unknown,
+            (op, Type::Instance(class)) => {
+                let class_member = class.class_member(self.db, op.dunder());
+                let call = class_member.call(self.db, &[operand_type]);
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2479]: If we get a not-callable outcome here, we should emit an unsupported-operation diagnostic (in the draft PR https://github.com/astral-sh/ruff/pull/13800 we are currently proposing to use the code `operation` for that diagnostic, though I'm kinda thinking I might prefer the more explicit `unsupported-operation` -- @AlexWaygood ?).

If you'd rather not implement that in this PR, that's fine, but please add a TODO comment for it here.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 7]
```diff
@@ -2518,9 +2527,11 @@ impl<'db> TypeInferenceBuilder<'db> {
                 .map(Type::IntLiteral)
                 .unwrap_or_else(|| KnownClass::Int.to_instance(self.db)),
 
-            (Type::IntLiteral(_), Type::IntLiteral(_), ast::Operator::Div) => {
-                KnownClass::Float.to_instance(self.db)
-            }
+            (
+                Type::IntLiteral(_),
+                Type::IntLiteral(_) | Type::BooleanLiteral(_),
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2532]: I wouldn't include `BooleanLiteral` here explicitly. (If you did, it should also be included for left operand, as well as right operand, so as to handle e.g. `True / 2`.) But I'd instead allow this to be handled by the more generic recursive-call approach described below.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 8]
```diff
@@ -2589,6 +2600,21 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+                | ruff_python_ast::Operator::Pow
+                | ruff_python_ast::Operator::LShift
+                | ruff_python_ast::Operator::RShift
+                | ruff_python_ast::Operator::BitXor
+                | ruff_python_ast::Operator::BitAnd
+                | ruff_python_ast::Operator::FloorDiv
+                | ruff_python_ast::Operator::MatMult => KnownClass::Int.to_instance(self.db),
+                ruff_python_ast::Operator::BitOr => KnownClass::Bool.to_instance(self.db),
+                ruff_python_ast::Operator::Div => KnownClass::Float.to_instance(self.db),
+            },
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2617]: `BitOr` of two `BooleanLiteral` needs special handling here; I'd give it its own case, and aim to return the correct `BooleanLiteral` for all four cases (should be easy to do as one line by just ORing the internal rust bools from each operand).

For everything else, I'd just have patterns for `(left, BooleanLiteral(b), op)` and for `(BooleanLiteral(b), right, op)` where all those cases do is construct the correct 1-or-0 `IntLiteral` for the `BooleanLiteral` (see above where that's already done in `infer_unary_expression`) and then re-run this match using that int literal in place of the boolean literal.

This will require extracting an `infer_binary_expression_type` function, which takes `left_ty: Type, right_ty: Type, op` arguments and just contains the match statement, so that you can recursively call it with the new types. You can see where we already did that in https://github.com/astral-sh/ruff/pull/13800, but it's fine to do it here too, one can easily rebase on the other.

File: crates/ruff_python_ast/src/nodes.rs [hunk 9]
```diff
@@ -2997,6 +2997,14 @@ impl UnaryOp {
             UnaryOp::USub => "-",
         }
     }
+    pub const fn dunder(self) -> &'static str {
+        match self {
+            UnaryOp::Invert => "__invert__",
+            UnaryOp::Not => "__bool__",
```
Comment (carljm) [inline crates/ruff_python_ast/src/nodes.rs:3003]: Hmm, the other three look good here, but this one can be backwards. Someone can write a class C with `def __bool__(self) -> Literal[True]: return True`, and now we'll think that `not C()` is `Literal[True]`, when it should be `Literal[False]`. (Let's add a test for this case!)

I think unfortunately this means that using a direct-translation `dunder` method here is probably not the right approach for unary ops, and instead this needs to be handled explicitly with a `match` over the op kinds in type inference, so that we can flip the sense for `not`.

**Review by carljm — CHANGES_REQUESTED** [review]
Thank you, this is great! A few comments:

File: crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md (hunk 1 above)
Comment (sharkdp) [reply crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md:24]: > a PR changing existing tests to use this style would also be welcome

I had already started with this, so here it is: https://github.com/astral-sh/ruff/pull/13806

File: crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md (hunk 2 above)
Comment (Glyphack) [reply crates/red_knot_python_semantic/resources/mdtest/binary/booleans.md]: I added more tests covering those cases.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 6 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs:2479]: > in the draft PR https://github.com/astral-sh/ruff/pull/13800 we are currently proposing to use the code `operation` for that diagnostic, though I'm kinda thinking I might prefer the more explicit `unsupported-operation` -- @AlexWaygood ?

We actually use the `operator` code in #13800. I don't mind using the `unsupported-` prefix, but I would much prefer to use the operat-_or_ suffix rather than operat-_ion_. To some extent any type checker diagnostic comes about as a result of an "unsupported operation"; but this error code specifically deals with attempts to use an operat-_or_ in an unsupported way.

TL;DR: `unsupported-operator`?

Comment (carljm) [reply crates/red_knot_python_semantic/src/types/infer.rs:2479]: Oops, my mistake; yeah `unsupported-operator` sounds good to me!

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 8 above)
Comment (Glyphack) [reply crates/red_knot_python_semantic/src/types/infer.rs:2617]: Thanks this definitely works better than what I did.

Just one note, I was thinking to also move the zero division check to the recursive call so if a non trivial zero division is covered in the recursions it's also reported but I found that cases like `True / False` will become `Cannot divide object of type `Literal[1]` by zero` so I left it in the parent function.
It makes sense because we can know anything that will become zero division error.

File: crates/ruff_python_ast/src/nodes.rs (hunk 9 above)
Comment (Glyphack) [reply crates/ruff_python_ast/src/nodes.rs:3003]: Thanks. I modified the code to to match other unary ops for instance types. We can also apply your previous suggestion here (match op and instance type and after getting the call type call the function again with return type and op) but I decided to let the Unary::Not case to handle all the Not operators to make it simpler to read.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 6 above)
Comment (Glyphack) [reply crates/red_knot_python_semantic/src/types/infer.rs:2484]: Thanks do we want to make functions to create the error with the right prefix in the future? I just put the string in the error message now.

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs:2484]: At some point we'll make our diagnostics less "stringly typed", and we'll probably do an audit of all our error codes at around the same time. For now I wouldn't much worry about :-)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 10]
```diff
@@ -2469,7 +2477,24 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+            (_, Type::Unknown) => Type::Unknown,
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), Type::Instance(class)) => {
+                let class_member = class.class_member(self.db, op.dunder());
+                let call = class_member.call(self.db, &[operand_type]);
+                if matches!(call, CallOutcome::NotCallable { not_callable_ty: _ }) {
+                    self.add_diagnostic(
+                        unary.into(),
+                        "unsupported-operator",
+                        format_args!(
+                            "Operator `{op}` is unsupported for type `{}`",
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2490]: Minor nit: since `+` and `-` are both unary and binary operators, let's make sure this diagnostic is clear which one we're talking about:
```suggestion
                            "Unary operator `{op}` is unsupported for type `{}`",
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 11]
```diff
@@ -2469,7 +2477,24 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+            (_, Type::Any) => Type::Any,
+            (_, Type::Unknown) => Type::Unknown,
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), Type::Instance(class)) => {
+                let class_member = class.class_member(self.db, op.dunder());
+                let call = class_member.call(self.db, &[operand_type]);
+                if matches!(call, CallOutcome::NotCallable { not_callable_ty: _ }) {
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2485]: There is also the possibility of `CallOutcome::Union` where some elements of the union were not callable, in which case we should also emit this diagnostic. The way to handle this is to match on `call.return_ty_result(db, unary, self)`; if we get `Ok(return_ty)` we just return that and there is no diagnostic, if we get any `NotCallableError::Err` variant, we do emit the unsupported-operator diagnostic, and return the `return_ty` field of the `Err` variant.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 12]
```diff
@@ -2484,17 +2509,32 @@ impl<'db> TypeInferenceBuilder<'db> {
...
         if matches!(
             (op, right_ty),
             (
                 ast::Operator::Div | ast::Operator::FloorDiv | ast::Operator::Mod,
-                Type::IntLiteral(0),
+                Type::IntLiteral(0)
             )
         ) {
-            self.check_division_by_zero(binary, left_ty);
+            // We want to print the original type that was used in the expression. The left type might
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2531]: I would prefer to leave this check out in `infer_binary_expression` and not re-check it recursively, and then we don't have to bother with `original_type` and this long comment. The only cost is that we have to explicitly check for `BooleanLiteral(False)` in the divisor, but that seems totally fine. It doesn't seem worth the complexity of adding `original_type` and needing this long comment, and the efficiency cost of double-checking for division by zero, just so we can match on `IntLiteral(0)` instead of `IntLiteral(0) | BooleanLiteral(False)`.

(It would be different if there were lots of different equivalent-to-zero values we might be converting to in a recursive call, but there aren't -- there's exactly one.)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 13]
```diff
@@ -2532,6 +2572,17 @@ impl<'db> TypeInferenceBuilder<'db> {
                 .map(Type::IntLiteral)
                 .unwrap_or_else(|| KnownClass::Int.to_instance(self.db)),
 
+            (Type::IntLiteral(n), Type::IntLiteral(m), ast::Operator::Pow) => {
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2575]: Let's add a few mdtests for `Pow` operator on int literals in `binary/integers.md`? (One for each possible case here: ok literal result, too large `m`, too large result.)

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 14]
```diff
@@ -2589,6 +2640,25 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+            (
+                Type::BooleanLiteral(b1),
+                Type::BooleanLiteral(b2),
+                ruff_python_ast::Operator::BitOr,
+            ) => match (b1, b2) {
+                (true, true) => Type::BooleanLiteral(true),
+                (true, false) => Type::BooleanLiteral(true),
+                (false, true) => Type::BooleanLiteral(true),
+                (false, false) => Type::BooleanLiteral(false),
+            },
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2652]: This can be simplified:
```suggestion
            ) => Type::BooleanLiteral(b1 | b2),
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 15]
```diff
@@ -2589,6 +2640,25 @@ impl<'db> TypeInferenceBuilder<'db> {
...
+
+            (Type::BooleanLiteral(_), right, op) => {
+                let int_value = left_ty.expect_int_literal();
+                self.infer_binary_expression_type(expr, op, Type::IntLiteral(int_value), right)
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2656]: Let's do it this way instead, like it's already done in `infer_unary_expression`:
```suggestion
            (Type::BooleanLiteral(bool_value), right, op) => {
                self.infer_binary_expression_type(expr, op, Type::IntLiteral(i64::from(bool_value)), right)
```
And same for the case below.

Bool and int literals aren't equivalent; the conversion between them should be explicit, it shouldn't be done implicitly by `expect_int_literal`.

File: crates/red_knot_python_semantic/src/types.rs [hunk 16]
```diff
@@ -355,6 +355,13 @@ impl<'db> Type<'db> {
     pub const fn into_int_literal_type(self) -> Option<i64> {
         match self {
             Type::IntLiteral(value) => Some(value),
+            Type::BooleanLiteral(b) => {
+                if b {
+                    Some(1)
+                } else {
+                    Some(0)
+                }
+            }
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:364]: We should not add this; boolean literals are not equivalent to int literals, the conversion should not be implicit. See below comment for how we can handle the conversion explicitly without doing this.

File: crates/ruff_python_ast/src/nodes.rs (hunk 9 above)
Comment (carljm) [reply crates/ruff_python_ast/src/nodes.rs:3003]: Sounds good! I think the results look fine now for this PR (it's fine for handling boolean value of instances to be a separate PR), but I still think we need to remove this `dunder()` method entirely. It's simply not correct that the unary Not operator is equivalent to the `__bool__` dunder, so leaving this method in place with that equivalence in it is asking for future bugs. Instead we should handle matching the `op` and getting the right dunder method for each op out in `infer_unary_expression`.

**Review by carljm — CHANGES_REQUESTED** [review]
Thanks for the updates! This now needs to be rebased now that #13800 has landed; there are some conflicts (should be minor.) And some comments below.

File: crates/ruff_python_ast/src/nodes.rs (hunk 9 above)
Comment (carljm) [reply crates/ruff_python_ast/src/nodes.rs:3003]: Or if you want to leave this method in place, it would have to return an `Option` and return `None` for unary `Not` operator. But at that point I'm not sure the ergonomics justify having the method at all, especially since I don't expect there to be multiple call sites where we need this.

Comment (Glyphack) [reply crates/ruff_python_ast/src/nodes.rs:3003]: No I'm removing it I agree with you.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 17]
```diff
@@ -2469,7 +2472,39 @@ impl<'db> TypeInferenceBuilder<'db> {
...
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+            (_, Type::Any) => Type::Any,
+            (_, Type::Unknown) => Type::Unknown,
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), Type::Instance(class)) => {
+                let unary_dunder_method = match op {
+                    UnaryOp::Invert => "__invert__",
+                    UnaryOp::UAdd => "__pos__",
+                    UnaryOp::USub => "__neg__",
+                    UnaryOp::Not => {
+                        debug_assert!(
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2483]: I think this should be `unreachable!` rather than `debug_assert!`

Comment (Glyphack) [reply crates/red_knot_python_semantic/src/types/infer.rs]: Oh I didn't know about this thanks.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 18]
```diff
@@ -2469,7 +2472,35 @@ impl<'db> TypeInferenceBuilder<'db> {
...
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+            (_, Type::Any) => Type::Any,
+            (_, Type::Unknown) => Type::Unknown,
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), Type::Instance(class)) => {
+                let unary_dunder_method = match op {
+                    UnaryOp::Invert => "__invert__",
+                    UnaryOp::UAdd => "__pos__",
+                    UnaryOp::USub => "__neg__",
+                    UnaryOp::Not => {
+                        unreachable!("Not operator is handled in it's own case");
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/infer.rs:2483]: Micro-nit (sorry!): "it's" is short for "it is"; this should be "its"

```suggestion
                        unreachable!("Not operator is handled in its own case");
```

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 17 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs]: No worries!

File: crates/red_knot_python_semantic/resources/mdtest/binary/integers.md [hunk 19]
```diff
@@ -11,6 +11,18 @@ reveal_type(-3 / 3)  # revealed: float
...
+## Power
+
+For power if the result fits in the int literal type it will be a Literal type. Otherwise the
+outcome is int.
+
+```py
+larget_u32 = 4_294_967_295
+reveal_type(2**2)  # revealed: Literal[4]
+reveal_type(1 ** (larget_u32 + 1))  # revealed: int
+reveal_type(2**larget_u32)  # revealed: int
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/binary/integers.md:23]:
```suggestion
largest_u32 = 4_294_967_295
reveal_type(2**2)  # revealed: Literal[4]
reveal_type(1 ** (largest_u32 + 1))  # revealed: int
reveal_type(2**largest_u32)  # revealed: int
```

File: crates/red_knot_python_semantic/resources/mdtest/binary/integers.md [hunk 20]
```diff
@@ -38,6 +50,12 @@ reveal_type(c)  # revealed: int
 # revealed: float
 reveal_type(int() / 0)
 
+# error: "Cannot divide object of type `Literal[1]` by zero"
+# revealed: float
+reveal_type(1 / False)
+True / False  # error: [division-by-zero] "Cannot divide object of type `Literal[True]` by zero"
+bool(1) / False  # error: [division-by-zero] "Cannot divide object of type `Literal[True]` by zero"
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/binary/integers.md:57]:
```suggestion
# error: [division-by-zero] "Cannot divide object of type `Literal[True]` by zero"
True / False
# error: [division-by-zero] "Cannot divide object of type `Literal[True]` by zero"
bool(1) / False
```

**Review by carljm — APPROVED** [review]
Looks great, thank you!!
